### ISSUE TITLE
Move Fl_Process class from FLUID to fltk.lib

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#678]
**Is your feature request related to a problem? Please describe.**
Fl_Process is a class in FLUID that runs shell commands. It can pipe the stdout of the shell command back to FLTK. This should be available as part of the main library because it would make FLTK a super easy solution for building front ends to scripts and shell commands.

**I am using FLTK**
 - Version 1.4.0

**Describe the solution you'd like**
I would like to move the code to the library, but make it into a widget. That way, Fl_Process is just another widget that can cause callbacks for the app to handle. The widget itself is a graphic representation of the state of the process. It doesn't need to be show or even linked to a parent to work. It should be easily linkable to an Fl_Simple_Terminal for debugging output. It should also provide a pipe back to the process's stdin.

**Comment by Albrecht-S** [comment]
I like the idea, although I don't know what **exactly** `Fl_Process` is doing. Need to look this up...

> ... move the code to the library, but make it into a widget ...

My first thought was: it should work w/o a specific widget (i.e. a GUI element) but then you wrote "it doesn't need to be show[n] or even linked to a parent to work". If this is the case, I think this would be fine. I'm thinking of users wanting to derive their own widget/GUI classes from `Fl_Process` to show the output or progress of a background process. This should be doable as well.

> It should be easily linkable to an Fl_[Simple_]Terminal for debugging output ...

That was indeed my next thought. I'd like to see Greg's comments to this particular feature. Note that he's working on replacing `Fl_Simple_Terminal` with his new `Fl_Terminal` which seems like a perfect fit because of its ability to handle escape sequences.

+1 :+1:

**Comment by ManoloFLTK** [comment]
What would be the public API of this new class? Of particular interest is how can FLTK receive a signal about new data availability in the pipe, since under Windows only sockets can be monitored via Fl::add_fd().

**Comment by erco77** [comment]
In my commercial applications I do this a lot across platforms.

I used an fltk timer and threads to manage the pipes, and an intermediate buffer between the parent and child threads to keep track of the data with locks.

Here's some details if it helps:

---

I called the class PipeCommand, as it envelopes running commands and reading them with pipes. There's separate unix+windows implementations (there was nothing different between linux and mac about doing this correctly).

For windows I used Job Objects to control processes running (CreateJobObject / AssignProcessToJobObject / TerminateJobObject / etc), so that hitting a button in FLTK could kill the running process cleanly (hard to do in Windows properly w/out job objects).

To start the process I use:
- separate CreatePipe()s for stdin + stdout
- assigned the stdout write pipe to manage stdout+stderr
-  CreateProcess() to start the process + assign it to a job object

A thread handles the dead loop that reads the pipe:
- while( WaitForSingleObject() ) to wait/loop for data to appear on the pipe
- ReadFile() to read the pipe, check for broken pipe/end of data/errors, exit thread
- Lock the data buffer, append the data, unlock
- continue while loop

The fltk timer polls the data buffer for new data, locks and appends the buffer to the widget, unlocks.

If the user clicks a button that involves cancelling the running child, it simply closes the class instance which first checks if the child is running, kills it with TerminateJobObject, kills the thread that watches the pipe with a mutex flag, and waits for it to finish, removes the timer, closes the pipes, and cleans up.

---

The devil's in the details, it's a lot of code.
The "data buffer" ended up being a small class of its own to manage appending and intercommunicating data between the child and main threads.

**Comment by MatthiasWM** [comment]
There are quite a bunch of sample implementations, one provided by Microsoft, but some better ones on Stack Overflow. It Is possible to have events attached for actions on any of the pipes that I can later capture using Fl::add_handler() or add_system_handler(). If that fails, @erco77 's suggestion to poll is possible too.

The graphics would be minimal, like on a modem: [oooo] where every circle represents a state. The first one is the process itself. off if unused, green when running, red if an error occurred. The remaining one show activity on stdin, stdout, and maybe stderr. Maybe a fifth one for signals sent, but that's probably overdoing it. Mind you, this widget will likely only be visible when debugging.

The api would be based on Fl_Box plus functions for sending data to stdin, and availability, peeking, and reading stdout and stderr. One command to set the command line, start, or stop the process. Lastly, there would need an enum to give callback reasons. Maybe:

```
class Fl_Process : Fl_Box {
public:
  // control the process
  void value(const char *command);
  int run(); // return error code
  int kill();
  // send to stdin, receive from stdout
  int send(const char *string);
  int avail();
  int peek(char *buffer, int n);
  int read(char *buffer, int n);
  int avail_stderr();
  int peek_stderr(char *buffer, int n);
  int read_stderr(char *buffer, int n);
  // give us a callback if we detect this prompt on stdout "
  void prompt(const char *string, double minimum_delay);
  // a bunch of callback reasons:
  const Fl_Callback_Reason FL_REASON_PROCESS_ENDED;
  const Fl_Callback_Reason FL_REASON_PROCESS_STDOUT_AVAIL;
  const Fl_Callback_Reason FL_REASON_PROCESS_STDERR_AVAIL;
  const Fl_Callback_Reason FL_REASON_PROCESS_PROMPT;
  // ability to link and unlink a terminal
  void terminal(Fl_Simple_Terminal *inout);
  // maybe signal handling
  const Fl_Callback_Reason FL_REASON_PROCESS_SIGNAL;
  void signal(int sig); // send a signal
  int signal(); // return the last signal received (fifo), or -1
};
```

I am sure there is more. Happy to hear suggestions.

**Comment by MatthiasWM** [comment]
> under Windows only sockets can be monitored via Fl::add_fd().

Yes, I think that this limitation comes from the problem that once we have a handle, we can not know if the handle was a pipe, socket, or whatever, so we agreed back then to assume it's a socket. But pretty much everything can be made to send events, only under MS Windows, you need different calls for every kind of stream.

If add_fd knew what type of handle we pass, we could implement that. The API for it is already relatively big though:
`void add_fd(int fd, int when, Fl_FD_Handler* cb, void* = 0);`, but adding an optional type may be tolerable?! I'd like to leave the existing API alone for now and stay away from eventually writing Fl_Pipe 🙄😂 .

**Comment by MatthiasWM** [comment]
> In my commercial applications I do this a lot across platforms. For windows I used Job Objects to control processes running (CreateJobObject / AssignProcessToJobObject / TerminateJobObject / etc), so that hitting a button in FLTK could kill the running process cleanly (hard to do in Windows properly w/out job objects).

Great pointer. Thank you!

> The devil's in the details, it's a lot of code.

Isn't it always, particularly on Windows?

**Comment by erco77** [comment]
In my class I provided two ways to read data from the pipe:
- Line oriented
- Block oriented
Line oriented was too slow for programs that spit out a LOT of info quickly, but line oriented was sometimes easier to deal with for parsing output.
I'd never needed "peek()", but there is a way to test if there's any output waiting or not.. basically the number of bytes waiting in the buffer; if it's non-zero there's stuff to be read for block style reading, and line oriented would simply remain 0 until at least a full line of data came in (at least one crlf in the buffer).
Regarding reading pipes with add_fd(), I suppose you could do what we do already on the mac; send a special even that triggers the callback whenever data has come in on the pipe.

**Comment by erco77** [comment]
Ya, it should merge just fine into Fl_Terminal. The test/terminal program I'm making uses popen() which works on unix but is sucky/annoying on windows (opens a DOS window if the app is windows). Something like this Fl_Process would be much better. There's some other example and cheat sheet apps that could make use of this too.

**Comment by MatthiasWM** [comment]
Not enough time before 1.4.0, but since it is an addition, it may go into 1.4.x

**Comment by erco77** [comment]
Ya, I think such a thing would have to be carefully designed and vetted to be "public".
I take it the initial need for this would be our example/test programs, and not so much for widgets, and perhaps secondarily providing user apps with a cross platform way to invoke processes a bit better than popen(), but perhaps not as crazy as e.g. the python subprocess library, which offers all kinds of stuff.

**Comment by MatthiasWM** [comment]
I see the most common use to write front ends for command line applications. For example, the pretty awesome TortoiseGIT on Windows is just a front end for whatever git command line version is installed. If done right (asynchronously with all three pipes and a way to kill processes), FLTK could make life really easy who have to added "something graphical" onto an existing shell tool.

**Comment by erco77** [comment]
Right -- I'd err on the side of looking at established stuff that has been "security vetted" to see what they do.

**Comment by MatthiasWM** [comment]
Not going to happen. Keep non-GUI stuff out of the core library.
